### PR TITLE
Two additional FixtureFactories

### PR DESCRIPTION
**Pull request by joelwetzel**
This PR adds two additional FixtureFactories for generating mock devices:

- Light Sensors
- Window Shades

### CODE REVIEW

**Comment by joelwetzel** [comment]
Hey, I wanted to give kudos on your framework.  With these new fixtureFactories and the toDateTime (in the other PR), I was just able to write some very readable tests, and found/fixed some bugs in one of my apps.

I have an app called Auto-Shades.  It uses a light sensor to automatically control a window shade, depending on the brightness at the window.  It also depends on time.  For example, if you manually override it, it's supposed to not fight back against you.  I've known there were some edge case bugs (because it didn't always behave as I predicted), but since it took a lot of time to test it on real devices, I was never able to track the bugs down.  Got them fixed now!

Anyway, I wanted to show you what the test cases can look like:

```
void "Shade should auto-close when it gets bright"() {
        given: "It's dark outside"
        shadeFixture.initialize(appExecutor, [windowShade: "open"])
        lightSensorFixture.initialize(appExecutor, [illuminance: 200])

        when: "The sun comes out"
        lightSensorFixture.setIlluminance(600)

        then: "The shade should auto-close"
        shadeFixture.state.windowShade == "closed"
        shadeFixture.state.position == 0
    }

void "Shade should not auto-close too soon after a manual open"() {
        given: "It's dark outside and the shades are open"
        shadeFixture.initialize(appExecutor, [windowShade: "open"])
        lightSensorFixture.initialize(appExecutor, [illuminance: 200])

        when: "A few minutes pass and the user manually closes the shade"
        appScript.addMinutes(5)
        shadeFixture.close()

        and: "Twenty minutes pass and it's still dark outside"
        appScript.addMinutes(20)
        lightSensorFixture.setIlluminance(200)

        then: "The shade should not auto-open"
        shadeFixture.state.windowShade == "closed"
        shadeFixture.state.position == 0
    }

    void "Shade can auto-open after a manual close, if enough time has passed"() {
        given: "It's dark outside and the shades are open"
        shadeFixture.initialize(appExecutor, [windowShade: "open"])
        lightSensorFixture.initialize(appExecutor, [illuminance: 200])

        when: "A few minutes pass and the user manually closes the shade"
        appScript.addMinutes(5)
        shadeFixture.close()

        and: "Two hours pass and it's still dark outside"
        appScript.addMinutes(121)
        lightSensorFixture.setIlluminance(200)

        then: "The shade should auto-open"
        shadeFixture.state.windowShade == "open"
        shadeFixture.state.position == 100
    }
```

**Comment by biocomp** [comment]
> Hey, I wanted to give kudos on your framework. With these new fixtureFactories and the toDateTime (in the other PR), I was just able to write some very readable tests, and found/fixed some bugs in one of my apps.

Thanks for sharing this, I'm really glad this is helpful to someone!

File: src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        def deviceInputValueFactory = new DeviceInputValueFactory([IlluminanceMeasurement])
+
+        def lightSensorDevice = deviceInputValueFactory.makeInputObject(name, 't',  DefaultAndUserValues.empty(), false)
+
+        def lightSensorMetaClass = lightSensorDevice.getMetaClass()
+
+        // Calling initialize attaches behavior involving commands, state maintenance, and sending events.
+        lightSensorMetaClass.initialize = { appExecutor, state ->
+            // The IlluminanceMeasurement capability has no commands and only a single state attribute: illuminance.
+            lightSensorMetaClass.state = state
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:26]: I vaguely remember that in groovy you can (probably?) tie actions to changes in properties. Can this be done to detect state changes instead of `setIlluminance` method?

File: src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+            // However, for the purposes of testing, I'm adding a public method to the fixture.
+            // It's not an actual command in real Hubitat, but it's useful for testing.
+            // It will let you set the illuminance state value, and publish an event for the change.
+            lightSensorMetaClass.setIlluminance = { int illuminance ->
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:31]: Remind me, does the light sensor change it state on receiving events? Do you need to handle incoming events?

**Review by biocomp — COMMENTED** [review]
Maybe add the factories to `hubitat_ci_example` examples and docs?

File: src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy (hunk 2 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:31]: No, light sensors don't receive any events from Hubitat.  The only thing a real device driver does is, when it receives an update from the physical device (over zwave or zigbee usually), it updates its illuminance value, and publishes an event.  (That hubitat apps can subscribe to.)

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:31]: Since we don't have a physical device, the setIlluminance method performs the work that would happen if a new value came in over Zigbee.

**Comment by joelwetzel** [comment]
> Maybe add the factories to `hubitat_ci_example` examples and docs?

I'll do that in a future update, once I have more factories done and the design has stabilized.

File: src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy (hunk 1 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:26]: I believe the setIlluminance method matches the convention:  https://stackoverflow.com/questions/52929624/intercept-property-setter-in-groovy

You could call lightSensorFixture.illuminance = 300, and it would set the value of lightSensorFixture.state.illuminance and then raise the event.

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:26]: But to intercept value changes of state.illuminance would require a smarter State object, instead of just a map.  And actually, Hubitat doesn't automatically raise events from state changes.  It requires the driver developer to set state and then send the event.

File: src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy (hunk 2 above)
Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:31]: I see. Thanks.

File: src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy (hunk 1 above)
Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/util/device_fixtures/LightSensorFixtureFactory.groovy:26]: Ok, thanks

**Review by biocomp — APPROVED** [review]

**Comment by joelwetzel** [comment]
I think we're ready to merge.  Thanks!
